### PR TITLE
Add dev environment

### PR DESCRIPTION
**Pull request by StephenCoady**
This PR creates a vagrant box to make working with fasjson easier.

Fasjson is running inside the vm along side a freeipa container and enrolled as a freeipa client.

It is still a work in progress as I need to clean up the task. 

This hasn't been tested fully so things might not be working.

**Verification**

From inside this repo just run:

`vagrant up` and both VMs should come up and be able to communicate.

**Note**: I needed to run `vagrant up --no-parallel ` as there is a problem with my vagrant install and it tried to install both VMs together - however this should not be the default behaviour and they should be provisioned sequentially.

**Requirements**
`vagrant`
plugins:
`vagrant-hostmanager`
`vagrant-hosts`
`vagrant-sshfs`

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -17,6 +17,6 @@
         "Topic :: Security",
         "Topic :: Software Development :: Libraries :: Python Modules",
     ],
-    install_requires=["flask", "python-ldap", "dnspython", "gssapi"],
+    install_requires=["flask", "python-ldap", "dnspython", "gssapi", "docker"],
```
Comment (tiran) [inline setup.py]: docker isn't a runtime dependency of fasjson. The RPM dependency generator will turn the requirement into a RPM package requirement, too.

Comment (StephenCoady) [reply setup.py]: woops. laziness brought that dep in and mindlessness meant it was left there :P

File: ansible/roles/fasjson/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+- name: Install python dependencies
+  shell: pip-3 install .
+  args:
+    chdir: /vagrant/
+
+- name: Copy wsgi
+  copy:
+    src: fasjson.wsgi
+    dest: /srv/
```
Comment (abompard) [inline ansible/roles/fasjson/tasks/main.yml]: I think the HTTPD configuration file looks for the wsgi script in `/usr/share/fasjson`, so I think it should be updated too.

File: ansible/roles/fasjson/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+- name: kinit
+  become: yes
+  shell: echo "adminPassw0rd!" | kinit admin
+    
+- name: Get service keytab for HTTPd
+  become: yes
+  shell: ipa-getkeytab -p HTTP/ipa.example.test -k /tmp/httpd.keytab
+    
+- name: Move keytab to gssproxy folder
```
Comment (abompard) [inline ansible/roles/fasjson/tasks/main.yml]: Why store the keytab in `/tmp` and then move it? Why not store it directly in the correct directory?

File: ansible/roles/fasjson/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+- name: Tune SELinux Policy
+  seboolean: 
+    name: httpd_can_connect_ldap
+    state: yes
+    persistent: yes
+
+- name: Create Apache service
+  shell: |
+    mkdir -p /etc/systemd/system/httpd.service.d
```
Comment (abompard) [inline ansible/roles/fasjson/tasks/main.yml]: You can do this with the `file` Ansible module.

File: ansible/roles/fasjson/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    mode: 0644
+    owner: root
+    group: root
+    
+- name: Enable and restart GSSProxy
+  systemd:
+    state: restarted
+    name: gssproxy
+    enabled: yes
+    daemon_reload: yes
```
Comment (abompard) [inline ansible/roles/fasjson/tasks/main.yml]: Why the daemon reload? I don't think we've changed the service file did we?

File: ansible/roles/fasjson/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    persistent: yes
+
+- name: Create Apache service
+  shell: |
+    mkdir -p /etc/systemd/system/httpd.service.d
+
+- name: Copy configs for Apache
+  copy: src={{ item.src }} dest={{ item.dest }}
+  with_items:
+    - { src: 'config/systemd-httpd-service-fasjson.conf', dest: '/etc/systemd/system/httpd.service.d/fasjson.conf' }
```
Comment (abompard) [inline ansible/roles/fasjson/tasks/main.yml]: I think you can use the Yaml dictionary syntax, like:
```yaml
- src: config/systemd-httpd-service-fasjson.conf
  dest: /etc/systemd/system/httpd.service.d/fasjson.conf
- src: config/httpd-fasjson.conf
  dest: /etc/httpd/conf.d/fasjson.conf
```

File: ansible/roles/freeipa/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+---
+- name: Install RPM packages
+  dnf:
+      name: ['git', 'vim', 'poetry', 'python3-flask', 'python3-pip', 'python3-cryptography', 'freeipa-server']
```
Comment (abompard) [inline ansible/roles/freeipa/tasks/main.yml]: Do we need all those RPMs there? Poetry and the other python libs don't seem necessary.

File: ansible/roles/fasjson/tasks/main.yml (hunk 5 above)
Comment (StephenCoady) [reply ansible/roles/fasjson/tasks/main.yml]: I'm not sure, we add a config but thats all. Its in the original instructions in the readme so thats why i added it here.

**Comment by abompard** [comment]
I'm getting the following error:
```
$ vagrant up --no-parallel
Bringing machine 'freeipa' up with 'libvirt' provider...
Bringing machine 'fasjson' up with 'libvirt' provider...
==> freeipa: Starting domain.
There was an error talking to Libvirt. The error message is shown
below:

Call to virDomainCreateWithFlags failed: Network not found: no network with matching name 'vagrant-private-dhcp'
```
Am I missing something?

File: ansible/roles/fasjson/tasks/main.yml (hunk 5 above)
Comment (abompard) [reply ansible/roles/fasjson/tasks/main.yml]: We didn't add a configuration for the service (in `/etc/systemd`) so I think we don't need it, but it's also not an issue to have it, no blocker here :-)

**Comment by StephenCoady** [comment]
@abompard thats strange. the only plugins I have installed are hostmanager, libvirt and sshfs. Unless you don't have hostmanager maybe? 

It looks like it might be something to do with your libvirt networks though. I will clean my environment and try again to see.

**Comment by StephenCoady** [comment]
@james02135 maybe could I also ask you to try this locally as you have the closest to a clean environment and we can see what happens?

**Comment by StephenCoady** [comment]
@abompard the problem was the unecessary ip assignment - it should be working now. you might need to install the hosts plugin. `vagrant plugin install vagrant-hosts`

**Comment by abompard** [comment]
OK! I now have a different error :-D
```
$ vagrant plugin install vagrant-hosts
Installing the 'vagrant-hosts' plugin. This can take a few minutes...
Vagrant failed to properly resolve required dependencies. These
errors can commonly be caused by misconfigured plugin installations
or transient network issues. The reported error is:

conflicting dependencies fog-core (~> 1.43.0) and fog-core (= 2.1.2)
  Activated fog-core-2.1.2
  which does not match conflicting dependency (~> 1.43.0)

  Conflicting dependency chains:
    fog-core (= 2.1.2), 2.1.2 activated

  versus:
    vagrant-libvirt (> 0), 0.0.41 activated, depends on
    fog-core (~> 1.43.0)
```

The fog-core 2.1.2 comes from the rubygem-fog-core package that F31 provides, and I can't remove it because it would pull vagrant-libvirt with it.

**Comment by StephenCoady** [comment]
@abompard this is a fairly long running fedora bug https://bugzilla.redhat.com/show_bug.cgi?id=1539181

For me anytime I need to add any plugin to vagrant I have to follow the steps below:

```
1. dnf remove vagrant-libvirt
2. vagrant plugin install $PLUGIN
3. dnf install vagrant-libvirt
```
Its because vagrant-libvirt is installed as a system plugin but vagrant won't work without the system rpm package :/

**Comment by abompard** [comment]
I'm still getting an error in the ipa-client-install step of the fasjson VM. If I run it directly I get:
```
# ipa-client-install --hostname ipa.example.test --domain example.test --realm EXAMPLE.TEST --server ipa.example.test -p admin -w adminPassw0rd! -U -N --force-join
This program will set up FreeIPA client.
Version 4.8.4

Client hostname: ipa.example.test
Realm: EXAMPLE.TEST
DNS Domain: example.test
IPA Server: ipa.example.test
BaseDN: dc=example,dc=test

Skipping chrony configuration
Successfully retrieved CA cert
    Subject:     CN=Certificate Authority,O=EXAMPLE.TEST
    Issuer:      CN=Certificate Authority,O=EXAMPLE.TEST
    Valid From:  2020-02-21 10:25:18
    Valid Until: 2040-02-21 10:25:18

Joining realm failed: HTTP POST to URL 'https://ipa.example.test:443/ipa/xml' failed.  HTTP response code is 401, not 200

Installation failed. Rolling back changes.
Disabling client Kerberos and LDAP configurations
nscd daemon is not installed, skip configuration
nslcd daemon is not installed, skip configuration
Client uninstall complete.
The ipa-client-install command failed. See /var/log/ipaclient-install.log for more information
```

And in the aforementioned log file I see:
```
2020-02-24T11:28:49Z DEBUG Starting external process
2020-02-24T11:28:49Z DEBUG args=['/usr/sbin/ipa-join', '-s', 'ipa.example.test', '-b', 'dc=example,dc=test', '-h', 'ipa.example.test', '-f']
2020-02-24T11:28:49Z DEBUG Process finished, return code=17
2020-02-24T11:28:49Z DEBUG stdout=
2020-02-24T11:28:49Z DEBUG stderr=HTTP POST to URL 'https://ipa.example.test:443/ipa/xml' failed.  HTTP response code is 401, not 200
2020-02-24T11:28:49Z ERROR Joining realm failed: HTTP POST to URL 'https://ipa.example.test:443/ipa/xml' failed.  HTTP response code is 401, not 200
2020-02-24T11:28:49Z ERROR Installation failed. Rolling back changes.
```

Any idea?

**Comment by StephenCoady** [comment]
@abompard the error doesn't tell much really so it doesn't, I wonder does the server tell us anything more useful. 

If you still have the freeipa box running - could you take a look at the contents of `/var/log/httpd/error_log`

I just my environment up and didn't get any error. I will investigate now and see what else I can find.

**Comment by odra** [comment]
I got it working by doing the following:

Removed all my libvirt packages and installed it again following those steps: https://docs.fedoraproject.org/en-US/quick-docs/getting-started-with-virtualization/#installing-virtualization-software

Install vagrant and vagrant-libvirt via dnf:

```
sudo dnf install -y vagrant vagrant-libvirt
```

Now this install a `fog-core` gem that uses a version which is not compatible with the other required plugins, so I needed to remove it and then install the correct version

```
sudo gem uninstall fog-core  -i  /usr/share/gems
sudo gem install fog-core -v 1.43.0
```

I was able to install the required plugins afterwards:

```
vagrant plugin install vagrant-host-manager vagrant-hosts vagrant-sshfs
```
I was then able to run `vagrant up -no-parallel`

Ansible is still running ATM so I will update my comment once it finishes. 

Note: I am using Fedora 31.

**Comment by StephenCoady** [comment]
Hi @tiran! I'm wondering if you had a few minutes to try this out? It seems to me like I have everything set up as described in the readme but I'm not sure if gssproxy is behaving as expected. I'm not familiar with this so if there was anything you could point to that you think is wrong in the box it would be a great help!

**Comment by abompard** [comment]
In the freeipa VM logs I get:
```
==> /var/log/krb5kdc.log <==
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): AS_REQ (7 etypes {aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), DEPRECATED:arcfour-hmac(23), camellia128-cts-cmac(25), camellia256-cts-cmac(26)}) 192.168.122.18: NEEDED_PREAUTH: admin@EXAMPLE.TEST for krbtgt/EXAMPLE.TEST@EXAMPLE.TEST, Additional pre-authentication required
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): closing down fd 11
Feb 25 14:18:59 ipa.example.test krb5kdc[2061](info): AS_REQ (7 etypes {aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), DEPRECATED:arcfour-hmac(23), camellia128-cts-cmac(25), camellia256-cts-cmac(26)}) 192.168.122.18: ISSUE: authtime 1582640339, etypes {rep=aes256-cts-hmac-sha1-96(18), tkt=aes256-cts-hmac-sha1-96(18), ses=aes256-cts-hmac-sha1-96(18)}, admin@EXAMPLE.TEST for krbtgt/EXAMPLE.TEST@EXAMPLE.TEST
Feb 25 14:18:59 ipa.example.test krb5kdc[2061](info): closing down fd 11
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): TGS_REQ (7 etypes {aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), DEPRECATED:arcfour-hmac(23), camellia128-cts-cmac(25), camellia256-cts-cmac(26)}) 192.168.122.18: ISSUE: authtime 1582640339, etypes {rep=aes256-cts-hmac-sha1-96(18), tkt=aes256-cts-hmac-sha1-96(18), ses=aes256-cts-hmac-sha1-96(18)}, admin@EXAMPLE.TEST for ldap/ipa.example.test@EXAMPLE.TEST
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): closing down fd 11
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): TGS_REQ (7 etypes {aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), DEPRECATED:arcfour-hmac(23), camellia128-cts-cmac(25), camellia256-cts-cmac(26)}) 192.168.122.18: ISSUE: authtime 1582640339, etypes {rep=aes256-cts-hmac-sha1-96(18), tkt=aes256-cts-hmac-sha1-96(18), ses=aes256-cts-hmac-sha1-96(18)}, admin@EXAMPLE.TEST for HTTP/ipa.example.test@EXAMPLE.TEST
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): closing down fd 11
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): TGS_REQ (1 etypes {aes256-cts-hmac-sha1-96(18)}) 192.168.122.18: ISSUE: authtime 1582640339, etypes {rep=aes256-cts-hmac-sha1-96(18), tkt=aes256-cts-hmac-sha1-96(18), ses=aes256-cts-hmac-sha1-96(18)}, admin@EXAMPLE.TEST for krbtgt/EXAMPLE.TEST@EXAMPLE.TEST
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): closing down fd 11
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): AS_REQ (7 etypes {aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), DEPRECATED:arcfour-hmac(23), camellia128-cts-cmac(25), camellia256-cts-cmac(26)}) 192.168.122.102: NEEDED_PREAUTH: HTTP/ipa.example.test@EXAMPLE.TEST for krbtgt/EXAMPLE.TEST@EXAMPLE.TEST, Additional pre-authentication required
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): closing down fd 11
Feb 25 14:18:59 ipa.example.test krb5kdc[2061](info): preauth (spake) verify failure: Preauthentication failed
Feb 25 14:18:59 ipa.example.test krb5kdc[2061](info): AS_REQ (7 etypes {aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), DEPRECATED:arcfour-hmac(23), camellia128-cts-cmac(25), camellia256-cts-cmac(26)}) 192.168.122.102: PREAUTH_FAILED: HTTP/ipa.example.test@EXAMPLE.TEST for krbtgt/EXAMPLE.TEST@EXAMPLE.TEST, Preauthentication failed
Feb 25 14:18:59 ipa.example.test krb5kdc[2061](info): closing down fd 11
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): AS_REQ (7 etypes {aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), DEPRECATED:arcfour-hmac(23), camellia128-cts-cmac(25), camellia256-cts-cmac(26)}) 192.168.122.102: NEEDED_PREAUTH: HTTP/ipa.example.test@EXAMPLE.TEST for krbtgt/EXAMPLE.TEST@EXAMPLE.TEST, Additional pre-authentication required
Feb 25 14:18:59 ipa.example.test krb5kdc[2062](info): closing down fd 11
Feb 25 14:18:59 ipa.example.test krb5kdc[2061](info): preauth (spake) verify failure: Preauthentication failed
Feb 25 14:18:59 ipa.example.test krb5kdc[2061](info): AS_REQ (7 etypes {aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), DEPRECATED:arcfour-hmac(23), camellia128-cts-cmac(25), camellia256-cts-cmac(26)}) 192.168.122.102: PREAUTH_FAILED: HTTP/ipa.example.test@EXAMPLE.TEST for krbtgt/EXAMPLE.TEST@EXAMPLE.TEST, Preauthentication failed
Feb 25 14:18:59 ipa.example.test krb5kdc[2061](info): closing down fd 11

==> /var/log/httpd/error_log <==
[Tue Feb 25 14:18:59.765966 2020] [auth_gssapi:error] [pid 2092:tid 139792986150656] [client 192.168.122.18:33960] GSS ERROR gss_acquire_cred[_from]() failed to get server creds: [Unspecified GSS failure.  Minor code may provide more information ( SPNEGO cannot find mechanisms to negotiate)], referer: https://ipa.example.test/ipa/xml

==> /var/log/httpd/access_log <==
192.168.122.18 - - [25/Feb/2020:14:18:59 +0000] "POST /ipa/xml HTTP/1.1" 401 2714
```

I have no idea what's going on. I'll destroy and rebuild the boxes and see if that helps.

File: ansible/roles/fasjson/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    src: "{{ item.src }}"
+    dest: "{{ item.dest }}"
+    mode: 0644
+    owner: vagrant
+    group: vagrant
+  with_items:
+    - src: "fasjson.wsgi"
+      dest: "/srv/"
+    - src: "fasjson.wsgi"
+      dest: "/usr/share/fasjson"
```
Comment (abompard) [inline ansible/roles/fasjson/tasks/main.yml]: Please choose one of these locations and adjust the HTTPd configuration accordingly.

File: ansible/roles/fasjson/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  args:
+    chdir: /vagrant/
+
+- name: Copy wsgi
+  copy: 
+    src: "{{ item.src }}"
+    dest: "{{ item.dest }}"
+    mode: 0644
+    owner: vagrant
+    group: vagrant
```
Comment (abompard) [inline ansible/roles/fasjson/tasks/main.yml]: You can add `selinux_context: httpd_sys_content_t` or HTTP will not see the file in `/srv`.

File: ansible/roles/fasjson/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+      dest: "/usr/share/fasjson"
+
+- name: Uninstall any old clients
+  shell: ipa-client-install --uninstall
+  ignore_errors: yes
+  become: yes
+  
+- name: Enroll system as IPA client
+  become: yes
+  shell: ipa-client-install --hostname ipa.example.test --domain example.test --realm EXAMPLE.TEST --server ipa.example.test -p admin -w adminPassw0rd! -U -N --force-join
```
Comment (abompard) [inline ansible/roles/fasjson/tasks/main.yml]: The `--hostname` should be `fasjson.example.test` I think

File: ansible/roles/fasjson/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+- name: kinit
+  become: yes
+  shell: echo "adminPassw0rd!" | kinit admin@EXAMPLE.TEST
+    
+- name: Get service keytab for HTTPd
+  become: yes
+  shell: ipa-getkeytab -p HTTP/ipa.example.test -k /var/lib/gssproxy/httpd.keytab
```
Comment (abompard) [inline ansible/roles/fasjson/tasks/main.yml]: The principal should be `HTTP/fasjson.example.test` I think

File: ansible/roles/fasjson/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    owner: vagrant
+    group: vagrant
+  with_items:
+    - src: "fasjson.wsgi"
+      dest: "/srv/"
+    - src: "fasjson.wsgi"
+      dest: "/usr/share/fasjson"
+
+- name: Uninstall any old clients
+  shell: ipa-client-install --uninstall
```
Comment (abompard) [inline ansible/roles/fasjson/tasks/main.yml]: You can also add `--unattended` so you can run `vagrant provision fasjson` and it won't wait for user input.

**Comment by abompard** [comment]
OK I *finally* found a way to make it work. We were basically missing an `ipa service-add` call. Here are the changes I needed to make:
```diff
--- a/ansible/roles/fasjson/tasks/main.yml
+++ b/ansible/roles/fasjson/tasks/main.yml
@@ -34,13 +34,25 @@
 - name: kinit
   become: yes
   shell: echo "adminPassw0rd!" | kinit admin@EXAMPLE.TEST
-    
+
+- name: Check that the HTTP service exists
+  shell: ipa service-find HTTP/fasjson.example.test
+  ignore_errors: True
+  register: ipa_service_exists
+
+- name: Create the service in IPA
+  become: yes
+  shell: ipa service-add HTTP/fasjson.example.test --force
+  when: ipa_service_exists.rc == 1
+
 - name: Get service keytab for HTTPd
   become: yes
-  shell: ipa-getkeytab -p HTTP/ipa.example.test -k /var/lib/gssproxy/httpd.keytab
-    
+  shell: ipa-getkeytab -p HTTP/fasjson.example.test -k /var/lib/gssproxy/httpd.keytab
+  args:
+    creates: /var/lib/gssproxy/httpd.keytab
+
 - name: Set the correct permissions on keytab
-  file: 
+  file:
     path: /var/lib/gssproxy/httpd.keytab
     owner: root
     group: root
```
Some of those changes are made to make the playbook idempotent (you can run it multiple times with `vagrant provision fasjson`).

**Pull request merged**